src/integrations/misc/readFileWithEncoding.ts

```
@@ -0,0 +1,153 @@
+	"js",
+	"jsx",
+	"ts",
+	"tsx",
+	"c",
+	"cpp",
+	"h",
+	"hpp",
+	"java",
+	"py",
+	"rb",
+	"php",
+	"go",
+	"rs",
+	"swift",
+	"scala",
+	"pl",
+	"lua",
+	"sh",
+	"bat",
+	"ps1",
+	"html",
+	"xml",
+	"css",
+	"scss",
+	"less",
+	"styl",
+	"vue",
+	"csv",
+	"tsv",
+]
+
+/**
+ * Scores text based on presence of Chinese/full-width characters
+ * Higher score means more likely to be Chinese text
+ * Pure ASCII text gets penalty score
+ */
+export function scoreText(text: string): number {
+	const total = text.length
+	if (total === 0) return 0
+	const zh = (text.match(/[\u4e00-\u9fff]/g) || []).length
+	const fullWidth = (text.match(/[\u3000-\u303F\uff00-\uffef]/g) || []).length
+	const ascii = (text.match(/[\x00-\x7F]/g) || []).length
+	return (zh * 2 + fullWidth) / total - (ascii === total ? 1 : 0)
+}
+
+/**
+ * Gets candidate encodings to try, prioritizing UTF-8 and Chinese-related encodings
+ * @param detected - Auto-detected encoding from chardet
+ */
+export function getCandidateEncodings(detected: string): string[] {
+	const baseEncodings = ["utf-8", "gb18030", "gbk", "shift_jis"]
+	const encSet = new Set(baseEncodings)
+	if (detected) encSet.add(detected.toLowerCase())
+	return Array.from(encSet)
+}
+
+/**
+ * Attempts to decode buffer with multiple encodings and returns best result
+ * @param buffer - File content buffer
+ * @param encodings - List of encodings to try
+ * @returns Object with decoded text, score and used encoding
+ */
+export function tryDecodeBuffer(
+	buffer: Buffer,
+	encodings: string[],
+): { text: string; score: number; encoding: string } {
+	let bestScore = -Infinity
+	let bestText = ""
+	let bestEncoding = "utf-8"
+
+	for (const enc of encodings) {
+		try {
+			const text = enc === "utf-8" || enc === "utf8" ? buffer.toString("utf8") : iconv.decode(buffer, enc)
+			const score = scoreText(text)
+			if (score > bestScore) {
+				bestScore = score
+				bestText = text
+				bestEncoding = enc
+			}
+		} catch {
+			continue
```

Should we maybe log these errors for debugging? Silent failures could make it hard to troubleshoot encoding issues later.